File: implementations/dotnet/src/Pcf/Crc.cs
namespace Pcf;

/// <summary>
/// Self-contained reflected-CRC routines for the three CRC algorithms in the
/// PCF hash registry (spec section 8.1). Each is parameterised by its reflected
/// polynomial, initial value and final XOR exactly as the specification states.
/// The bitwise form is intentionally simple and auditable; the test suite pins
/// each variant to its canonical check value for the input "123456789".
/// </summary>
internal static class Crc
{
    /// <summary>CRC-32/ISO-HDLC (the CRC used by zlib, gzip and PNG).</summary>
    public static uint Crc32IsoHdlc(byte[] data) =>
        Compute32(data, 0xEDB8_8320u, 0xFFFF_FFFFu, 0xFFFF_FFFFu);

    /// <summary>CRC-32C (Castagnoli).</summary>
    public static uint Crc32c(byte[] data) =>
        Compute32(data, 0x82F6_3B78u, 0xFFFF_FFFFu, 0xFFFF_FFFFu);

    /// <summary>
    /// CRC-64/XZ (as used by xz/liblzma and System.IO.Hashing.Crc64). The
    /// bitwise right-shift form below takes the <em>reflected</em> polynomial
    /// 0xC96C5795D7870F42, i.e. the bit-reversal of the spec's stated
    /// 0x42F0E1EBA9EA3693.
    /// </summary>
    public static ulong Crc64Xz(byte[] data) =>
        Compute64(data, 0xC96C_5795_D787_0F42ul, 0xFFFF_FFFF_FFFF_FFFFul, 0xFFFF_FFFF_FFFF_FFFFul);

    private static uint Compute32(byte[] data, uint poly, uint init, uint xorout)
    {
        uint crc = init;
        foreach (byte b in data)
        {
            crc ^= b;
            for (int i = 0; i < 8; i++)
            {
                crc = (crc & 1) != 0 ? (crc >> 1) ^ poly : crc >> 1;
            }
        }
        return crc ^ xorout;
    }

    private static ulong Compute64(byte[] data, ulong poly, ulong init, ulong xorout)
    {
        ulong crc = init;
        foreach (byte b in data)
        {
            crc ^= b;
            for (int i = 0; i < 8; i++)
            {
                crc = (crc & 1) != 0 ? (crc >> 1) ^ poly : crc >> 1;
            }
        }
        return crc ^ xorout;
    }
}

File: implementations/ts/pcf/src/crc.ts
/**
 * Pure-TypeScript CRC implementations used by the PCF hash registry
 * (spec section 8.1). `@noble/hashes` does not provide CRCs, so the three
 * registered variants are implemented here from their normative parameters.
 *
 * All three are *reflected* CRCs (refin = refout = true), so the table-driven
 * form processes the low byte and shifts right, using the reflected polynomial.
 */

const MASK32 = 0xffff_ffff;
const MASK64 = (1n << 64n) - 1n;

function makeTable32(reflectedPoly: number): Uint32Array {
  const table = new Uint32Array(256);
  for (let n = 0; n < 256; n++) {
    let c = n;
    for (let k = 0; k < 8; k++) {
      c = c & 1 ? reflectedPoly ^ (c >>> 1) : c >>> 1;
    }
    table[n] = c >>> 0;
  }
  return table;
}

function makeTable64(reflectedPoly: bigint): BigUint64Array {
  const table = new BigUint64Array(256);
  for (let n = 0n; n < 256n; n++) {
    let c = n;
    for (let k = 0; k < 8; k++) {
      c = c & 1n ? reflectedPoly ^ (c >> 1n) : c >> 1n;
    }
    table[Number(n)] = c & MASK64;
  }
  return table;
}

// Reflected polynomials for each registered CRC (spec section 8.1).
const CRC32_TABLE = makeTable32(0xedb88320); // CRC-32/ISO-HDLC
const CRC32C_TABLE = makeTable32(0x82f63b78); // CRC-32C (Castagnoli)
// CRC-64/XZ: normal poly 0x42F0E1EBA9EA3693, reflected 0xC96C5795D7870F42.
const CRC64_TABLE = makeTable64(0xc96c5795d7870f42n);

function crc32With(table: Uint32Array, data: Uint8Array): number {
  let crc = MASK32;
  for (let i = 0; i < data.length; i++) {
    crc = (crc >>> 8) ^ table[(crc ^ data[i]!) & 0xff]!;
  }
  return (crc ^ MASK32) >>> 0;
}

/** CRC-32/ISO-HDLC (the CRC used by zlib, gzip, and PNG). */
export function crc32(data: Uint8Array): number {
  return crc32With(CRC32_TABLE, data);
}

/** CRC-32C (Castagnoli). */
export function crc32c(data: Uint8Array): number {
  return crc32With(CRC32C_TABLE, data);
}

/** CRC-64/XZ. */
export function crc64(data: Uint8Array): bigint {
  let crc = MASK64;
  for (let i = 0; i < data.length; i++) {
    const idx = Number((crc ^ BigInt(data[i]!)) & 0xffn);
    crc = (crc >> 8n) ^ CRC64_TABLE[idx]!;
  }
  return (crc ^ MASK64) & MASK64;
}
